crates/ide-diagnostics/src/handlers/external_lints.rs

```diff
@@ -0,0 +1,156 @@
+use std::{cell::RefCell, path::PathBuf};
+
+use ide_db::base_db::FileId;
+use linter_adapter::context::{DriverContext, DriverContextWrapper};
+use linter_api::{
+    ast::{
+        item::{self, CommonItemData, ItemId, ItemType, ModItem, StaticItem},
+        ty::Mutability,
+        BodyId, Crate, CrateId, Span, Symbol,
+    },
+    context::AstContext,
+};
+use once_cell::unsync::OnceCell;
+use syntax::{
+    ast::{HasModuleItem, HasVisibility, Item},
+    SourceFile, SyntaxNode, TextRange, TextSize, AstNode,
+};
+
+use crate::{Diagnostic, DiagnosticsConfig};
+
+struct RADriver<'a> {
+    diags: RefCell<&'a mut Vec<Diagnostic>>,
+    id_map: RefCell<&'a mut Vec<SyntaxNode>>,
+    ast_context: OnceCell<&'a AstContext<'a>>,
+}
+
+fn leak<'a, T: 'a>(x: T) -> &'a T {
+    Box::leak(Box::new(x))
+}
+
+impl<'ast> DriverContext<'ast> for RADriver<'ast> {
+    fn emit_lint(
+        &self,
```

Comment:
Is it intentionally `&self`? By changing it to `&mut self` I can drop those `RefCell` fields in `RADriver`.

Reply:
Good question, this is mostly the result of evolving source code. Just using `&self` was also inspired by rustc way to implement these traits.

The main question here is how concurrency should be handled. By only using `&self` the adapter passes responsibility to the driver. The driver could potentially check multiple crates at once. On the other hand, we could just use `RefCell` in the adapter itself :thinking: